### PR TITLE
fix: fullnode sync restart

### PR DESCRIPTION
**Pull request by purusang**
- l2 block finalization is done only if the l2 block actually exists.
- check point finalization is triggered only once in events L1Block or NewTipBlock.

### Reason of failing to restart
When fullnode is in initial syncing state, l1 reader/processor reads/processes blocks way faster than l2 block reader/processor. As a result l2blocks that did not exist in l2 db were marked as finalized. 

After restart, fullnode tries to perform state transitions from the point where l2 block was last finalized. If l2 block does not exist it crashes.

### Solution
In this pr I have changed the behavior of how we actually mark the finalize block. Previously, we were updating finalized block based on checkpoints processed by reading l1 blocks only. 
After this pr, we do as previously only if the l2 block actually exist in the l2 db. And for each new l2 block tip we look for any corresponding checkpoint and finalize it if the l2blockid in checkpoint matches.

### No effect on Sequencer
Sequencer does not get in this situation so it's work should not be affected.

### No effect on fully synced nodes
If fullnode is in sync with realtime chain state then no way it receives any l1 block which will have future l2 checkpoint.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Refactor
- [ ] New or updated tests
- [ ] Dependency Update

## Notes to Reviewers

I want to add it's test case in a separate pr as creating this rare condition is quite tricky in test cases. Any suggestions on test cases for this condition will be highly appreciated.
I have tested these changes against our devnet releases/0.1.0 and the fix must be applied asap to releases.
## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code where necessary.
- [x] I have updated the documentation if needed.
- [x] My changes do not introduce new warnings.
- [ ] I have added tests that prove my changes are effective or that my feature works.
- [x] New and existing tests pass with my changes.

## Related Issues
STR-647

### CODE REVIEW

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 1]
```diff
@@ -261,6 +285,52 @@ fn handle_maturable_height(
...
+        }
+        None => {
+            debug!(%blkid, "L2 block not found in verified checkpoints, possibly not a last block in the checkpoint.");
+        }
+    }
+    (writes, actions)
+}
+
+/// Searches for a given `L2BlockId` within a slice of `L1Checkpoint` structs
+/// and returns the height of the corresponding L1 block if found.
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:323]:
```suggestion
/// Searches for a given [`L2BlockId`] within a slice of [`L1Checkpoint`] structs
/// and returns the height of the corresponding L1 block if found.
```

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 2]
```diff
@@ -261,6 +285,52 @@ fn handle_maturable_height(
...
+    checkpoints: &[L1Checkpoint],
+    target_l2_blockid: &L2BlockId,
+) -> Option<u64> {
+    checkpoints
+        .binary_search_by(|checkpoint| checkpoint.batch_info.l2_blockid.cmp(target_l2_blockid))
+        .ok()
+        .map(|index| checkpoints[index].height)
+}
+
 /// Filters a list of `BatchCheckpoint`s, returning only those that form a valid sequence
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:334]:
```suggestion
/// Filters a list of [`BatchCheckpoint`]s, returning only those that form a valid sequence
```

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 3]
```diff
@@ -261,6 +285,52 @@ fn handle_maturable_height(
     (writes, actions)
 }
 
+fn handle_checkpoint_finalization(
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:288]: Can you add docstrings here?

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 4]
```diff
@@ -224,15 +224,20 @@ pub fn process_event<D: Database>(
...
+            writes.extend(wrs);
+            actions.extend(acts);
         }
     }
 
     Ok(ClientUpdateOutput::new(writes, actions))
 }
 
-fn handle_maturable_height(
+fn handle_mature_l1_height(
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:237]: Add docstrings here?

**Review by storopoli — CHANGES_REQUESTED** [review]
Some nits on docstrings.
The logic looks fine, but I am not a codeowner in this part of the codebase.

**Review by delbonis — DISMISSED** [review]
We'll be coming back around to working on some of this later.

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 5]
```diff
@@ -224,15 +224,41 @@ pub fn process_event<D: Database>(
...
+///
+/// # Arguments
+///
+/// * `maturable_height` - The height at which L1 blocks are considered mature.
+/// * `state` - A reference to the current client state.
+/// * `database` - A reference to the database interface.
+///
+/// # Returns
+///
+/// A tuple containing:
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:255]:
```suggestion
/// A tuple containing:
///
```

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 6]
```diff
@@ -261,7 +306,74 @@ fn handle_maturable_height(
...
+/// # Arguments
+///
+/// * `state` - A reference to the current client state.
+/// * `blkid` - A reference to the L2 block ID to be finalized.
+/// * `params` - A reference to the rollup parameters.
+/// * `database` - A reference to the database interface.
+///
+/// # Returns
+///
+/// A tuple containing:
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:327]:
```suggestion
/// A tuple containing:
///
```

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 7]
```diff
@@ -261,7 +306,74 @@ fn handle_maturable_height(
...
+            }
+        }
+        None => {
+            debug!(%blkid, "L2 block not found in verified checkpoints, possibly not a last block in the checkpoint.");
+        }
+    }
+    (writes, actions)
+}
+
+/// Searches for a given [`L2BlockId`] within a slice of [`L1Checkpoint`] structs
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:364]:
```suggestion
/// Searches for a given [`L2BlockId`] within an array of [`L1Checkpoint`]
```

File: crates/consensus-logic/src/csm/client_transition.rs [hunk 8]
```diff
@@ -270,12 +382,13 @@ fn handle_maturable_height(
 /// # Arguments
 ///
 /// * `state` - The client's current state, which provides the L1 view and pending checkpoints.
-/// * `checkpoints` - A slice of `BatchCheckpoint`s to be filtered.
+/// * `checkpoints` - A slice of [`BatchCheckpoint`]s to be filtered.
```
Comment (storopoli) [inline crates/consensus-logic/src/csm/client_transition.rs:385]:
```suggestion
/// * `checkpoints` - An array of [`BatchCheckpoint`]s to be filtered.
```

**Review by storopoli — CHANGES_REQUESTED** [review]
Amazing docstrings.
Just some suggestions.
I still think "slice" is less intuitive than "array". But feel free to disagree and go with "slice".

File: crates/consensus-logic/src/csm/client_transition.rs (hunk 8 above)
Comment (purusang) [reply crates/consensus-logic/src/csm/client_transition.rs:385]: Yeah for reading arrays are more intuitive but these are slices so keeping as it it.

**Review by storopoli — APPROVED** [review]
ACK c58fac0218e7536834be6235e57fac5e620d12e8

**Review by delbonis — APPROVED** [review]

**Pull request merged**